core/capabilities/webapi/outgoing_connector_handler.go

```diff
@@ -0,0 +1,223 @@
+// TODO: handle retries and timeouts
+func (c *OutgoingConnectorHandler) HandleSingleNodeRequest(ctx context.Context, messageID string, payload []byte) (*api.Message, error) {
+	ch := make(chan *api.Message, 1)
+	c.responseChsMu.Lock()
+	c.responseChs[messageID] = ch
+	c.responseChsMu.Unlock()
+	l := logger.With(c.lggr, "messageID", messageID)
+	l.Debugw("sending request to gateway")
+
+	body := &api.MessageBody{
+		MessageId: messageID,
+		DonId:     c.gc.DonID(),
+		Method:    c.method,
+		Payload:   payload,
+	}
+
+	// simply, send request to first available gateway node from sorted list
+	// this allows for deterministic selection of gateway node receiver for easier debugging
+	gatewayIDs := c.gc.GatewayIDs()
+	if len(gatewayIDs) == 0 {
+		return nil, errors.New("no gateway nodes available")
+	}
+	sort.Strings(gatewayIDs)
+
+	err := c.gc.SignAndSendToGateway(ctx, gatewayIDs[0], body)
+	if err != nil {
+		return nil, errors.Wrap(err, "failed to send request to gateway")
+	}
+
+	select {
+	case resp := <-ch:
+		return resp, nil
+	case <-ctx.Done():
+		return nil, ctx.Err()
+	}
+}
+
+func (c *OutgoingConnectorHandler) HandleGatewayMessage(ctx context.Context, gatewayID string, msg *api.Message) {
+	body := &msg.Body
+	l := logger.With(c.lggr, "gatewayID", gatewayID, "method", body.Method, "messageID", msg.Body.MessageId)
+	if !c.rateLimiter.Allow(body.Sender) {
+		// error is logged here instead of warning because if a message from gateway is rate-limited,
+		// the workflow will eventually fail with timeout as there are no retries in place yet
+		c.lggr.Errorw("request rate-limited")
+		return
+	}
+	l.Debugw("handling gateway request")
+	switch body.Method {
+	case capabilities.MethodWebAPITarget:
+		var payload capabilities.TargetResponsePayload
+		unmarshalAndSendResponse(ctx, msg, payload, c, l)
+	case capabilities.MethodComputeAction:
+		var payload sdk.FetchResponse
+		unmarshalAndSendResponse(ctx, msg, payload, c, l)
+	default:
+		l.Errorw("unsupported method")
+	}
+}
+
+func unmarshalAndSendResponse[T any](ctx context.Context, msg *api.Message, payload T, c *OutgoingConnectorHandler, l logger.Logger) {
+	body := &msg.Body
+	err := json.Unmarshal(body.Payload, &payload)
+	if err != nil {
+		l.Errorw("failed to unmarshal payload", "err", err)
+		return
+	}
+	c.responseChsMu.Lock()
+	defer c.responseChsMu.Unlock()
+	ch, ok := c.responseChs[body.MessageId]
+	if !ok {
+		l.Errorw("no response channel found")
+		return
+	}
+	select {
+	case ch <- msg:
+		delete(c.responseChs, body.MessageId)
+	case <-ctx.Done():
+		return
+	}
+}
+
+func (c *OutgoingConnectorHandler) Start(ctx context.Context) error {
+	return c.gc.AddHandler([]string{c.method}, c)
+}
+
+func (c *OutgoingConnectorHandler) Close() error {
+	return nil
+}
+
+func (c *OutgoingConnectorHandler) CreateFetcher(workflowID, workflowExecutionID string) func(req *wasmpb.FetchRequest) (*wasmpb.FetchResponse, error) {
```

Comment:
This isn't great from a coupling point of view since you're mixing two concepts together (fetcher + connector handler): Can we move this into the compute capability as a wrapper type?

Reply:
moved! tbh I don't see this approach wrong given this is basically a wrapper around the connector. Actually the first implementation was in compute (as an inline function) and we talked about moving it as a wrapper of OutgoingConnectorHandler [here](https://github.com/smartcontractkit/chainlink/pull/14723#discussion_r1796678661). In any case I've done the move in an isolated commit so we can keep any of the versions

Reply:
I think we misunderstood one another: creating the OutgoingConnectorHandler was definitely the right move -- when I was talking about a wrapper I meant something that lives in the compute capability and looks like eg.

```
func newFetchFactory(connector) *fetchFactory {
    ... your constructor
}

func (*fetcherFactory) NewFetcher() {
    ... this is where you do your wrapping
}
```

The reason why I'm not wild about adding this to the connector is that the concept of a fetcher wouldn't be useful elsewhere -- it's a concept that's specific to custom compute and the wrapping of it should therefore live there.